File: bench/prime.c
/* prime.c — Count primes up to N (C reference implementation for the benchmark)
   Compile: gcc -O2 -o prime prime.c
   Usage:   ./prime --n 500000
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int is_prime(int n) {
    if (n < 2) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    for (int i = 3; i * i <= n; i += 2)
        if (n % i == 0) return 0;
    return 1;
}

static int count_primes(int limit) {
    int count = 0;
    for (int num = 2; num <= limit; num++)
        if (is_prime(num)) count++;
    return count;
}

int main(int argc, char *argv[]) {
    int n = 500000;
    for (int i = 1; i < argc - 1; i++)
        if (strcmp(argv[i], "--n") == 0)
            n = atoi(argv[i + 1]);

    printf("Primes up to %d: %d\n", n, count_primes(n));
    return 0;
}

File: bench/prime.py
"""prime.py — Count primes up to N (Python reference implementation for the benchmark)
Usage: python3 prime.py --n 500000
"""
import sys


def is_prime(n: int) -> bool:
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True


def count_primes(limit: int) -> int:
    return sum(1 for num in range(2, limit + 1) if is_prime(num))


def main() -> None:
    n = 500000
    args = sys.argv[1:]
    for i, arg in enumerate(args):
        if arg == "--n" and i + 1 < len(args):
            n = int(args[i + 1])

    result = count_primes(n)
    print(f"Primes up to {n}: {result}")


if __name__ == "__main__":
    main()
